### ISSUE TITLE
UnicodeDecodeError with Django Rest Framework on header parsing

### ISSUE THREAD

**Issue by fredericcambon** [evonove/django-oauth-toolkit#202]
How to reproduce :
- Add `OAuth2Authentication` on top of your `authentication_classes`
- Call one of your endpoint in Javascript with a header such as {'Content-Type': 'lål'} (Any non-ascii char should do)

```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py", line 111, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/django/views/decorators/csrf.py", line 57, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/viewsets.py", line 85, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/django/views/decorators/csrf.py", line 57, in wrapped_view
    return view_func(*args, **kwargs)
  File "##################/api/views.py", line 155, in dispatch
    response = super(BaseAPIViewSet, self).dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/views.py", line 407, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/views.py", line 395, in dispatch
    self.initial(request, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/views.py", line 324, in initial
    self.perform_authentication(request)
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/views.py", line 274, in perform_authentication
    request.user
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/request.py", line 271, in user
    self._authenticate()
  File "/usr/local/lib/python2.7/dist-packages/rest_framework/request.py", line 457, in _authenticate
    user_auth_tuple = authenticator.authenticate(self)
  File "/usr/local/lib/python2.7/dist-packages/oauth2_provider/ext/rest_framework/authentication.py", line 18, in authenticate
    valid, r = oauthlib_core.verify_request(request, scopes=[])
  File "/usr/local/lib/python2.7/dist-packages/oauth2_provider/oauth2_backends.py", line 135, in verify_request
    valid, r = self.server.verify_request(uri, http_method, body, headers, scopes=scopes)
  File "/usr/local/lib/python2.7/dist-packages/oauthlib/oauth2/rfc6749/endpoints/base.py", line 64, in wrapper
    return f(endpoint, uri, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/oauthlib/oauth2/rfc6749/endpoints/resource.py", line 68, in verify_request
    request = Request(uri, http_method, body, headers)
  File "/usr/local/lib/python2.7/dist-packages/oauthlib/common.py", line 389, in __init__
    self.headers = CaseInsensitiveDict(encode(headers or {}))
  File "/usr/local/lib/python2.7/dist-packages/oauthlib/common.py", line 385, in <lambda>
    encode = lambda x: to_unicode(x, encoding) if encoding else x
  File "/usr/local/lib/python2.7/dist-packages/oauthlib/common.py", line 331, in to_unicode
    return dict(((to_unicode(k, encoding), to_unicode(v, encoding)) for k, v in data))
  File "/usr/local/lib/python2.7/dist-packages/oauthlib/common.py", line 331, in <genexpr>
    return dict(((to_unicode(k, encoding), to_unicode(v, encoding)) for k, v in data))
  File "/usr/local/lib/python2.7/dist-packages/oauthlib/common.py", line 317, in to_unicode
    return unicode_type(data, encoding=encoding)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xf1 in position 10: unexpected end of data
```

DRF tries to match all `authentication_classes` to the request until it finds the right one. Since `OAuth2Authentication` is the first one in the list in our case, it is tested first.
The code fails when trying to decode the headers of the request

Now I know that the exception is thrown in the underneath library `oauthlib` but I wasn't exactly sure if it is a problem that could be solved directly in `django-oauth-toolkit`

**Comment by boris-savic** [comment]
Can confirm this is still an issue today

**Comment by jleclanche** [comment]
This appears to be fixed in oauthlib 2.0 and Python 3 so I'm closing this.

There is still a crash if Content-Type contains unicode characters with DRF but that's not an issue in DOT.
